sdlib/d/gc/slab.d

```diff
@@ -23,6 +25,159 @@ struct SlabAllocGeometry {
 	}
 }
 
+struct SlabAllocInfo {
+private:
+	SlabAllocGeometry sg;
+	Extent* e;
+	bool allowsMetaData = false;
+
+	void setMetaData(bool value) {
+		assert(allowsMetaData, "size class not supports slab metadata!");
+
+		if (hasMetaData == value) {
+			return;
+		}
+
+		hasMetaData = value;
+		if (hasMetaData) {
+			e.enableMetaData(sg.index);
+			return;
+		}
+
+		e.disableMetaData(sg.index);
+	}
+
+	@property
+	T* ptrToAllocEnd(T)() {
+		return cast(T*) (sg.address + sg.size) - T.sizeof;
+	}
+
+	alias freeSpacePtr = ptrToAllocEnd!ushort;
+
+	@property
+	size_t freeSpace() {
+		return hasMetaData ? readPackedFreeSpace(freeSpacePtr) : 0;
+	}
+
+	void setFreeSpace(size_t size) {
+		assert(allowsMetaData, "size class not supports slab metadata!");
+
+		if (size == 0) {
+			setMetaData(false);
+			return;
+		}
+
+		writePackedFreeSpace(freeSpacePtr, size & ushort.max);
+		setMetaData(true);
+	}
```

Comment:
Don't have a ton of private feature on top. This not what anyone wants to read first.